### PR TITLE
KAFKA-17465: Make getMembersFromGroup be non-blocking

### PR DESCRIPTION
**Pull request by TaiJuWu**
Jira: https://issues.apache.org/jira/browse/KAFKA-17465

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 1]
```diff
@@ -4114,27 +4114,30 @@ private Integer nodeFor(ConfigResource resource) {
...
+    private KafkaFutureImpl<List<MemberIdentity>> getMembersFromGroup(String groupId, String reason) {
+        KafkaFutureImpl<List<MemberIdentity>> future = new KafkaFutureImpl<>();
 
-        List<MemberIdentity> membersToRemove = new ArrayList<>();
-        for (final MemberDescription member : members) {
-            MemberIdentity memberIdentity = new MemberIdentity().setReason(reason);
+        describeConsumerGroups(Collections.singleton(groupId)).describedGroups().get(groupId).whenComplete((res, ex) -> {
+            if (ex != null) {
+                future.completeExceptionally(ex);
+            }
```
Comment (frankvicky) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Do we need to keep executing following logic even if we encounter exception ?
In original one, it seems that the exception will be wrapped as `KafkaException` and throw.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 2]
```diff
@@ -4114,27 +4114,30 @@ private Integer nodeFor(ConfigResource resource) {
...
+            Collection<MemberDescription> members = res.members();
+            List<MemberIdentity> membersToRemove = new ArrayList<>();
+            for (final MemberDescription member : members) {
```
Comment (frankvicky) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: `for (final MemberDescription member : res.members())`

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 3]
```diff
@@ -4114,27 +4114,30 @@ private Integer nodeFor(ConfigResource resource) {
...
 
-            if (member.groupInstanceId().isPresent()) {
-                memberIdentity.setGroupInstanceId(member.groupInstanceId().get());
-            } else {
-                memberIdentity.setMemberId(member.consumerId());
+                if (member.groupInstanceId().isPresent()) {
+                    memberIdentity.setGroupInstanceId(member.groupInstanceId().get());
+                } else {
+                    memberIdentity.setMemberId(member.consumerId());
+                }
```
Comment (frankvicky) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: nit: How about leverage api of `Optional` ?
```java
  member.groupInstanceId()
          .map(memberIdentity::setGroupInstanceId)
          .orElseGet(() -> memberIdentity.setMemberId(member.consumerId()));
```

**Review by frankvicky — COMMENTED** [review]
Hi @TaiJuWu 
I have few comments, PTAL

**Comment by frankvicky** [comment]
@TaiJuWu 
Please fix the PR title, correct JIRA number is `17465`

**Comment by TaiJuWu** [comment]
> @TaiJuWu Please fix the PR title, correct JIRA number is `17465`

Thanks for your remainder, I will address your comments later.

**Comment by TaiJuWu** [comment]
> Hi @TaiJuWu I have few comments, PTAL

Hi @frankvicky ,thanks for review, all comments are addressed.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 4]
```diff
@@ -4114,27 +4114,27 @@ private Integer nodeFor(ConfigResource resource) {
...
+    private KafkaFutureImpl<List<MemberIdentity>> getMembersFromGroup(String groupId, String reason) {
+        KafkaFutureImpl<List<MemberIdentity>> future = new KafkaFutureImpl<>();
 
-        List<MemberIdentity> membersToRemove = new ArrayList<>();
-        for (final MemberDescription member : members) {
-            MemberIdentity memberIdentity = new MemberIdentity().setReason(reason);
+        describeConsumerGroups(Collections.singleton(groupId)).describedGroups().get(groupId).whenComplete((res, ex) -> {
+            if (ex != null) {
+                future.completeExceptionally(new KafkaException("Encounter exception when trying to get members from group: " + groupId, ex));
+            }
```
Comment (frankvicky) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: My question is:
In the original logic, if an exception is caught, the method stops executing and throws the exception. However, in your PR, the handling process seems to differ from this.

Comment (TaiJuWu) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Thanks, you are right😺

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 5]
```diff
@@ -4143,19 +4144,26 @@ public RemoveMembersFromConsumerGroupResult removeMembersFromConsumerGroup(Strin
...
-        List<MemberIdentity> members;
-        if (options.removeAll()) {
-            members = getMembersFromGroup(groupId, reason);
-        } else {
-            members = options.members().stream()
-                .map(m -> m.toMemberIdentity().setReason(reason))
-                .collect(Collectors.toList());
+        final SimpleAdminApiFuture<CoordinatorKey, Map<MemberIdentity, Errors>> future =
+                RemoveMembersFromConsumerGroupHandler.newFuture(groupId);
+        if (!options.removeAll()) {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Could you please try to reduce duplicate code?

Comment (TaiJuWu) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Done. Thanks for review!

**Review by frankvicky — APPROVED** [review]
LGTM

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 6]
```diff
@@ -4143,22 +4144,33 @@ public RemoveMembersFromConsumerGroupResult removeMembersFromConsumerGroup(Strin
...
+                    .collect(Collectors.toList());
+            handleRemoveMembersFromConsumerGroupAndInvokeDriver(groupId, members, options.timeoutMs, future);
+            return new RemoveMembersFromConsumerGroupResult(future.get(CoordinatorKey.byGroupId(groupId)), options.members());
         }
 
-        SimpleAdminApiFuture<CoordinatorKey, Map<MemberIdentity, Errors>> future =
-                RemoveMembersFromConsumerGroupHandler.newFuture(groupId);
-        RemoveMembersFromConsumerGroupHandler handler = new RemoveMembersFromConsumerGroupHandler(groupId, members, logContext);
-        invokeDriver(handler, future, options.timeoutMs);
+        getMembersFromGroup(groupId, reason).whenComplete((members, ex) -> {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: how about:
```java
        KafkaFutureImpl<List<MemberIdentity>> f;
        if (options.removeAll()) f = getMembersFromGroup(groupId, reason);
        else {
            f = new KafkaFutureImpl<>();
            f.complete(options.members().stream()
                    .map(m -> m.toMemberIdentity().setReason(reason))
                    .collect(Collectors.toList()));
        }

        f.whenComplete((members, ex) -> {
            if (ex != null) {
                future.completeExceptionally(Collections.singletonMap(CoordinatorKey.byGroupId(groupId), ex));
            } else {
                RemoveMembersFromConsumerGroupHandler handler = new RemoveMembersFromConsumerGroupHandler(groupId, members, logContext);
                invokeDriver(handler, future, options.timeoutMs());
            }
        });
```

Comment (TaiJuWu) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Done. Thanks for review.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 7]
```diff
@@ -4114,27 +4114,28 @@ private Integer nodeFor(ConfigResource resource) {
...
+        KafkaFutureImpl<List<MemberIdentity>> future = new KafkaFutureImpl<>();
 
-            if (member.groupInstanceId().isPresent()) {
-                memberIdentity.setGroupInstanceId(member.groupInstanceId().get());
+        describeConsumerGroups(Collections.singleton(groupId)).describedGroups().get(groupId).whenComplete((res, ex) -> {
+            if (ex != null) {
+                future.completeExceptionally(new KafkaException("Encounter exception when trying to get members from group: " + groupId, ex));
             } else {
-                memberIdentity.setMemberId(member.consumerId());
+                List<MemberIdentity> membersToRemove = new ArrayList<>();
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Could you please initialize the size by `res.members()`?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 8]
```diff
@@ -4114,27 +4114,28 @@ private Integer nodeFor(ConfigResource resource) {
...
+                for (final MemberDescription member : res.members()) {
+                    MemberIdentity memberIdentity = new MemberIdentity().setReason(reason);
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: If you prefer to use lambda function, please try to eliminate the local variable. for example:`
```java
                    membersToRemove.add(member.groupInstanceId()
                            .map(id -> new MemberIdentity().setGroupInstanceId(id))
                            .orElseGet(() -> new MemberIdentity().setMemberId(member.consumerId()))
                            .setReason(reason));
```

**Review by chia7712 — COMMENTED** [review]
@TaiJuWu thanks for your updates. a couple of comments are left.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 8 above)
Comment (TaiJuWu) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Hi @chia7712 , thanks for review.
I rewrite this part, please take a look.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 9]
```diff
@@ -4143,19 +4140,27 @@ public RemoveMembersFromConsumerGroupResult removeMembersFromConsumerGroup(Strin
...
-            members = getMembersFromGroup(groupId, reason);
-        } else {
-            members = options.members().stream()
-                .map(m -> m.toMemberIdentity().setReason(reason))
-                .collect(Collectors.toList());
+        final SimpleAdminApiFuture<CoordinatorKey, Map<MemberIdentity, Errors>> future =
+                RemoveMembersFromConsumerGroupHandler.newFuture(groupId);
+
+        KafkaFutureImpl<List<MemberIdentity>> f;
+        if (options.removeAll()) f = getMembersFromGroup(groupId, reason);
```
Comment (frankvicky) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: nit: I would recommend not omitting the curly braces, or at the very least, it should break the line.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 10]
```diff
@@ -4143,19 +4140,27 @@ public RemoveMembersFromConsumerGroupResult removeMembersFromConsumerGroup(Strin
...
-        if (options.removeAll()) {
-            members = getMembersFromGroup(groupId, reason);
-        } else {
-            members = options.members().stream()
-                .map(m -> m.toMemberIdentity().setReason(reason))
-                .collect(Collectors.toList());
+        final SimpleAdminApiFuture<CoordinatorKey, Map<MemberIdentity, Errors>> future =
+                RemoveMembersFromConsumerGroupHandler.newFuture(groupId);
+
+        KafkaFutureImpl<List<MemberIdentity>> f;
```
Comment (frankvicky) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Please rename this variable.

**Review by frankvicky — COMMENTED** [review]
Thanks for update.
I have some comments.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 10 above)
Comment (TaiJuWu) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: All comments are addressed, thanks!

**Review by chia7712 — APPROVED** [review]

**Pull request merged**